pkg/orchestrator/controller/controller_test.go

```diff
@@ -0,0 +1,409 @@
+package controller
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/testing"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/submariner-io/admiral/pkg/federate/mocks"
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	submarinerClientset "github.com/submariner-io/submariner/pkg/client/clientset/versioned"
+	fakeSubm "github.com/submariner-io/submariner/pkg/client/clientset/versioned/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/klog"
+)
+
+type watchReactor struct {
+	endpointsWatchStarted chan bool
+	clustersWatchStarted  chan bool
+}
+
+type eqCluster struct {
+	expected *submarinerv1.Cluster
+}
+
+type eqEndpoint struct {
+	expected *submarinerv1.Endpoint
+}
+
+var _ = Describe("Submariner Orchestrator", func() {
+	klog.InitFlags(nil)
+
+	When("start is called", testStart)
+	Describe("Cluster lifecycle notifications", testClusterLifecycleNotifications)
+	Describe("Submariner resource distribution", testResourceDistribution)
+})
+
+func testStart() {
+	var (
+		mockCtrl      *gomock.Controller
+		mockFederator *mocks.MockFederator
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		mockFederator = mocks.NewMockFederator(mockCtrl)
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
```

Comment:
nit: I'm a proponent of leaving the state as it was before the test. So here I'd nil both variables too.

(same for the other tests)

Reply:
sure although it's not really necessary as each should be initialized in `BeforeEach`.

Reply:
I'm aware. Just stating my preference :)